### ISSUE TITLE
Issue with get_metrics()

### ISSUE THREAD

**Issue by ashys2012** [baal-org/baal#298]
This is the first time I am raising an issue on Github, so aplogies for any mistakes. 

The issue is, when I try to print the metrics after train_on_dataset I am getting a value for test accuracy as well identical copy of train accuracy. Similarly, the issue is also with test_on_dataset.  Below I have copy pasted the code as well as the output.

Kaggle notebook for furhter ref: https://www.kaggle.com/code/anthonyashwin/bayesian-al-neu-steel-ds

**Code**

**Debug function to print metric states**
def print_metric_states(metrics, stage):
    print(f"Metrics during {stage}:")
    for k, v in metrics.items():
        if hasattr(v, 'value'):
            print(f"{k}: {v.value}")
        elif hasattr(v, 'compute'):
            print(f"{k}: {v.compute()}")

 **Add metrics to the model**
baal_model.add_metric(name='accuracy', initializer=lambda: accuracy_metric)
baal_model.add_metric(name='f1', initializer=lambda: f1_metric)
baal_model.add_metric("ece", initializer=lambda: ece_metric)
baal_model.add_metric("ece_cls", initializer=lambda: ece_cls_metric)

**1. Train the model for a few epochs on the training set.**
baal_model.train_on_dataset(active_learning_ds, optimizer, batch_size=16, epoch=10, use_cuda=USE_CUDA)
print("after training")
print_metric_states(baal_model.metrics, "training")

**Print training metrics**
training_metrics = baal_model.get_metrics()
print(training_metrics)

**2. Reset metrics before testing**
baal_model._reset_metrics()
print("after resetting metrics")
print_metric_states(baal_model.metrics, "after resetting")

**3. Evaluate the model on the test set**
baal_model.test_on_dataset(test_dataset, batch_size=16, use_cuda=USE_CUDA)
print("after testing")
print_metric_states(baal_model.metrics, "testing")

**Print testing metrics**
testing_metrics = baal_model.get_metrics()
print(testing_metrics)

**Output:**

[27196-MainThread] [baal.modelwrapper:train_on_dataset:86] 2024-07-20T04:49:01.807920Z [info     ] Starting training              dataset=400 epoch=10
[27196-MainThread] [baal.modelwrapper:train_on_dataset:97] 2024-07-20T04:49:28.432936Z [info     ] Training complete              train_loss=1.4844413995742798
after training
Metrics during training:
test_loss: 1.7386467456817627
train_loss: 1.4844413995742798
**test_accuracy: 0.3675000071525574
train_accuracy: 0.3675000071525574
test_f1: 0.3675000071525574
train_f1: 0.3675000071525574
test_ece: 0.02902940109372137
train_ece: 0.02902940109372137**
test_ece_cls: [0.18125317 0.24410557 0.3045224  0.17053657 0.25430272 0.56281751]
train_ece_cls: [0.18125317 0.24410557 0.3045224  0.17053657 0.25430272 0.56281751]
{'test_loss': 1.7386467456817627, 'train_loss': 1.4844413995742798, 'test_accuracy': 0.3675000071525574, 'train_accuracy': 0.3675000071525574, 'test_f1': 0.3675000071525574, 'train_f1': 0.3675000071525574, 'test_ece': 0.02902940109372137, 'train_ece': 0.02902940109372137, 'test_ece_cls': array([0.18125317, 0.24410557, 0.3045224 , 0.17053657, 0.25430272,
       0.56281751]), 'train_ece_cls': array([0.18125317, 0.24410557, 0.3045224 , 0.17053657, 0.25430272,
       0.56281751]), 'dataset_size': 400}
after resetting metrics
Metrics during after resetting:
test_loss: nan
train_loss: nan
test_accuracy: nan
train_accuracy: nan
test_f1: 0.0
train_f1: 0.0
test_ece: nan
train_ece: nan
test_ece_cls: [0. 0. 0. 0. 0. 0.]
train_ece_cls: [0. 0. 0. 0. 0. 0.]
[27196-MainThread] [baal.modelwrapper:test_on_dataset:126] 2024-07-20T04:49:28.447607Z [info     ] Starting evaluating            dataset=414
[/home/achazhoor/anaconda3/envs/al_cl/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36](https://file+.vscode-resource.vscode-cdn.net/home/achazhoor/anaconda3/envs/al_cl/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:36): UserWarning: The ``compute`` method of metric F1Score was called before the ``update`` method which may lead to errors, as metric states have not yet been updated.
  warnings.warn(*args, **kwargs)
[/home/achazhoor/anaconda3/envs/al_cl/lib/python3.8/site-packages/baal/utils/metrics.py:142](https://file+.vscode-resource.vscode-cdn.net/home/achazhoor/anaconda3/envs/al_cl/lib/python3.8/site-packages/baal/utils/metrics.py:142): RuntimeWarning: invalid value encountered in divide
  return ((self.samples [/](https://file+.vscode-resource.vscode-cdn.net/) n) * np.abs(self._acc() - average_confidence)).sum()
[27196-MainThread] [baal.modelwrapper:test_on_dataset:136] 2024-07-20T04:49:29.281374Z [info     ] Evaluation complete            test_loss=1.3887609243392944
after testing
Metrics during testing:
test_loss: 1.3887609243392944
train_loss: nan
**test_accuracy: 0.4546703100204468
train_accuracy: 0.4546703100204468
test_f1: 0.45410627126693726
train_f1: 0.45410627126693726
test_ece: 0.1364660037384517
train_ece: 0.1364660037384517**
test_ece_cls: [0.11939948 0.1041922  0.34043689 0.26438838 0.25740336 0.60956178]
train_ece_cls: [0.11939948 0.1041922  0.34043689 0.26438838 0.25740336 0.60956178]
{'test_loss': 1.3887609243392944, 'train_loss': nan, 'test_accuracy': 0.4546703100204468, 'train_accuracy': 0.4546703100204468, 'test_f1': 0.45410627126693726, 'train_f1': 0.45410627126693726, 'test_ece': 0.1364660037384517, 'train_ece': 0.1364660037384517, 'test_ece_cls': array([0.11939948, 0.1041922 , 0.34043689, 0.26438838, 0.25740336,
       0.60956178]), 'train_ece_cls': array([0.11939948, 0.1041922 , 0.34043689, 0.26438838, 0.25740336,
       0.60956178]), 'dataset_size': 400}

**Expected behavior**
I expected to see different values for training and testing accuracy. I feel so the issue might be from intitalization of the specific metrics. For loss metrics, there is no issue. 

**Vesion (please complete the following information):**
 - OS: Ubuntu
 - Python: 3.7
 - Baal version: 2.0.0

**Comment by Dref360** [comment]
Hello! Thank for submitting the issue.

I have trouble accessing the notebook, it says there are no saved version.

In any case, I suspect that `baal_model.add_metric(name='accuracy', initializer=lambda: accuracy_metric)` is the issue. Can you post what is `accuracy metric`? 

Metric object keep track of the values internally so if you provide the same object to both `test` and `train` it will give you this behavior.

See the example [here](https://github.com/baal-org/baal/blob/eb123cfec9e81a6a88410049bc3836376ba0e68c/docs/api/utils.md?plain=1#L21)

Everytime `initializer` is called it returns a new object Accuracy.

```python
wrapper.add_metric(name='accuracy',initializer=lambda : Accuracy())
```

**Comment by Dref360** [comment]
Adding a warning in ModelWrapper could be handy if someone wants to make the contribution.

**Comment by ashys2012** [comment]
The below code worked for me. The output prints out NAN values which is alright since, after resetting the metrics, it should be NAN. Better than printing incorrect values. Also, for F1score, we need to explicitly move it into device, since it was initialized by torchmetrics!

baal_model.add_metric('accuracy', lambda: Accuracy())
baal_model.add_metric('f1', lambda : F1Score().to(device))
baal_model.add_metric("ece", lambda: ECE())
baal_model.add_metric("ece_per_cls", lambda: ECE_PerCLs(num_classes))

# 1. Train the model for a few epochs on the training set.
baal_model.train_on_dataset(active_learning_ds, optimizer, batch_size=16, epoch=10, use_cuda=USE_CUDA)
print("after training")
print_metric_states(baal_model.metrics, "training")

# Print training metrics
training_metrics = baal_model.get_metrics()
print(training_metrics)

# 2. Reset metrics before testing
baal_model._reset_metrics()
print("after resetting metrics")
print_metric_states(baal_model.metrics, "after resetting")

# 3. Evaluate the model on the test set
baal_model.test_on_dataset(test_dataset, batch_size=16, use_cuda=USE_CUDA)
print("after testing")
print_metric_states(baal_model.metrics, "testing")


**The output:**

[58782-MainThread] [baal.modelwrapper:train_on_dataset:86] 2024-07-20T06:07:12.237550Z [info     ] Starting training              dataset=400 epoch=10
[58782-MainThread] [baal.modelwrapper:train_on_dataset:97] 2024-07-20T06:07:39.058920Z [info     ] Training complete              train_loss=1.4986889362335205
after training
Metrics during training:
test_loss: 1.6059362888336182
train_loss: 1.4986889362335205
test_accuracy: nan
train_accuracy: 0.3449999988079071
test_f1: 0.0
train_f1: 0.3449999988079071
test_ece: nan
train_ece: 0.039717854149639624
test_ece_per_cls: [0. 0. 0. 0. 0. 0.]
train_ece_per_cls: [0.20245503 0.23051898 0.26758713 0.10565668 0.22037682 0.52136208]
{'test_loss': 1.6059362888336182, 'train_loss': 1.4986889362335205, 'test_accuracy': nan, 'train_accuracy': 0.3449999988079071, 'test_f1': 0.0, 'train_f1': 0.3449999988079071, 'test_ece': nan, 'train_ece': 0.039717854149639624, 'test_ece_per_cls': array([0., 0., 0., 0., 0., 0.]), 'train_ece_per_cls': array([0.20245503, 0.23051898, 0.26758713, 0.10565668, 0.22037682,  0.52136208]), 'dataset_size': 400}

Metrics after resetting:
test_loss: nan
train_loss: nan
test_accuracy: nan
train_accuracy: nan
test_f1: 0.0
train_f1: 0.0
test_ece: nan
train_ece: nan
test_ece_per_cls: [0. 0. 0. 0. 0. 0.]
train_ece_per_cls: [0. 0. 0. 0. 0. 0.]
[58782-MainThread] [baal.modelwrapper:test_on_dataset:126] 2024-07-20T06:07:39.064312Z [info     ] Starting evaluating            dataset=414
[58782-MainThread] [baal.modelwrapper:test_on_dataset:136] 2024-07-20T06:07:39.897241Z [info     ] Evaluation complete            test_loss=1.5070700645446777
after testing
Metrics during testing:
test_loss: 1.5070700645446777
train_loss: nan
test_accuracy: 0.3794642984867096
train_accuracy: nan
test_f1: 0.37922704219818115
train_f1: 0.0
test_ece: 0.10316963123094632
train_ece: nan
test_ece_per_cls: [0.24156527 0.21219999 0.42159821 0.30897063 0.20933304 0.70937823]
train_ece_per_cls: [0. 0. 0. 0. 0. 0.]
